### DISCUSSION TITLE
Text Wrapping in Label Node

### DISCUSSION THREAD

**Discussion by noman-khan1224** [eclipse-glsp/glsp#1333] [Q&A]
We're working on incorporating text wrapping into our labelNode, using the GLabelView to create it, which typically produces a <text> tag. Given that our text content might exceed the fixed size of the RectangularNode in our setup, is there a method to enable text wrapping for nodes with longer text content?

**Comment by planger** [comment]
SVG has no builtin support for line wrapping. As a workaround you can use a `ForeignObjectElement` (see [example in docs](https://eclipse.dev/glsp/documentation/rendering/#foreignobjectviewhttpsgithubcomeclipsesprottyblobmasterpackagessprottysrclibgeneric-viewstsx)) and re-use the text wrapping capability of the browser. Note that the rendering will then only work in browsers though.

I believe @rsoika has a SVG-native implementation for a similar requirement though.

**Comment by rsoika** [comment]
Hi @noman-khan1224 , yes we have implemented a custom view on the client, that splits up a longer label into multiple lines. We use `<tspan>` tags as these tags are also supported in a SVG export. We compute the layout completely on the client view. So the server just set the element type.

Find the implementation here: https://github.com/imixs/open-bpmn/blob/48a21aeebc511a39a69b55e8d02eeba8d5605528/open-bpmn.glsp-client/open-bpmn-glsp/src/bpmn-element-views.tsx#L538

And you can try it out in our online demo here:

https://open-bpmn.imixs.org/#/usr/src/app/open-bpmn.glsp-client/workspace

The labels of an event node and the Text Annotation are supporting this feature.

**Reply by noman-khan1224** [reply]
Hi @rsoika,

I have tried your approach. It's working fine until the length of word is less than than the width of node. Here what I am aiming to achieve is to wrap the text according to node-width similar to TextArea in Html.

![image](https://github.com/eclipse-glsp/glsp/assets/161299971/172a9e13-f5e6-4663-aa0a-f005c69fbec8)

**Reply by rsoika** [reply]
Hi @noman-khan1224 , ok than take a look at the method textLineSplitter https://github.com/imixs/open-bpmn/blob/48a21aeebc511a39a69b55e8d02eeba8d5605528/open-bpmn.glsp-client/open-bpmn-glsp/src/bpmn-element-views.tsx#L579C10-L579C26
You can adapt this method to your needs.

**Reply by arcanefoam** [reply]
I did a small change to the bpmn implementation to improve the width calculation, as it resulted in lines that left too much space to the right. It uses a canvas object to calculate the text width based on the font used.
```ts
    // re-use canvas object for better performance
    private readonly canvas = document.createElement('div');

    /**
     * This method splits a text into an array of separate lines.
     * Beside the new-line character also the length of a line is considered.
     * This is needed to display a long text in multiple SVG tspan elements.
     *
     * @param text - The text to be split.
     * @returns an array of text lines
     */
    private textLineSplitter(text: string, nodeWidth: number): string[] {
        const result: string[] = [];
        // first split only by newlines
        const lines = text.split(/[\n\r]+/);
        // Next we iterate of each line and verify if the words
        // in this line fit into the given nodesWidth
        for (let i = 0; i < lines.length; i++) {
            // split line into words, only using horizontal whitespace
            const words = lines[i].split(/[^\S\n\r]+/);
            // now lets add word by word and verify if we need to add a newline...
            let line = '';
            for (let j = 0; j < words.length; j++) {
                const word = words[j];
                // Test if adding the next word would exceed the node width
                const lineWidth = this.getTextWidth(line + word + ' ');
                if (lineWidth > nodeWidth) {
                    result.push(line);
                    line = word + ' ';
                } else {
                    line += word + ' ';
                }
            }
            result.push(line);
        }
        return result;
    }

    /**
     * Uses canvas.measureText to compute and return the width of the given text of given font in pixels.
     *
     * @see https://stackoverflow.com/questions/118241/calculate-text-width-with-javascript/21015393#21015393
     */
    private getTextWidth(text: string) {
        this.canvas.textContent = text;
        this.canvas.style.position = 'absolute';
        this.canvas.style.float = 'left';
        this.canvas.style.whiteSpace = 'nowrap';
        this.canvas.style.visibility = 'hidden';
        this.canvas.style.font = MultiLineTextNodeView.FONT;
        document.body.appendChild(this.canvas);
        const width = this.canvas.clientWidth;
        document.body.removeChild(this.canvas);
        return width;
    }
```

**Comment by arcanefoam** [comment]
Sorry for reviving this discussion. I implemented the approach from bpmn suggested by @rsoika and works like a charm. The issue I am having is that I would like the text to be editable. If I change  MultiLineTextNode(GShapeElement) to be an SLabelImpl (EditableLabel) I can not decrease the width of the node once I make it wider.  In my case the MultiLine is nested inside another node.

I've been trying to debug the code to find the issue and I believe it has something to do with VBoxLayouterExt and how it calculates the width based on its children... but I can't figure out why it doe snot allow the width to me smaller than the current width.

**Reply by rsoika** [reply]
Hi @arcanefoam , I know that the implementation in Open-BPMN is not perfect as we use a fixed character width. But at the moment it is working. I added a new issue https://github.com/imixs/open-bpmn/issues/442 so we can improve this part in the future. Thanks for the input!
